docs/src/main/asciidoc/extension-maturity-model.adoc

```diff
@@ -0,0 +1,93 @@
+
+[id="extension-maturity-model"]
+= A maturity model for Quarkus extensions
+include::_attributes.adoc[]
+:diataxis-type: concept
+:categories: writing-extensions
+:topics: extensions
+:summary: Quarkus extensions can do a lot, or a little. This guide explains some of the capabilities extension authors may wish to include.
+////
+The document header ends at the first blank line. Do not remove the blank line between the header and the abstract summary.
+////
+
+What makes a good Quarkus extension? What capabilities is a Quarkus extension expected to provide?
+
+* It actually works
+** Works in JVM mode
+** Works in dev mode
+** Works in native
+* Dev service (if there is an external service dependency)
+* Developer joy
+** Configuration support
+** Use build-time application knowledge to remove boilerplate
+* Dev UI
+* Supersonic subatomic performance
+** Use build-time application knowledge to eliminate wasteful runtime code paths
+* Codestart application template
+
+The order in this model isn't exact. Different developers will have different views on what capabilities are most important. You may wish to (say) prioritise performance over enhancing your extension's Dev UI tile. That's fine!
+Also, not every step will apply to every extension. For example, you don't need a dev service if your extension doesn't depend on external services.
+
+Also note that this list only includes the technical features of your extension.
+You might also want to think about how you share your extension.
+The link:https://hub.quarkiverse.io/checklistfornewprojects/[new extension checklist] on the Quarkiverse Hub has a useful list of ways extensions can participate in the ecosystem.
+
+Here are some pointers on how to achieve those capabilities.
+
+== It actually works
+
+=== Works in JVM mode
+
+For most extensions, this is the minimum expectation.
+When wrapping an existing library, this is usually trivial to achieve; if an extension is providing net-new capability, it might be a bit more work. Quarkus provides tools for xref:writing-extensions.adoc#testing-extensions[unit testing and integration testing] extensions.
+
+=== Works in dev mode
+
+In some cases, extra work may be needed to ensure any wrapped libraries can tolerate
```

Comment:
good point but wouldn't it be better fitting on dev joy?

Reply:
See the comment above - I think if we're suggesting an implementation order, we can't group too much. So let's get rid of 'Developer Joy' altogether from that list, so there's no question of things being sub-points under it or not. 

Developer Joy should be a theme, not an implementation order.

Reply:
i see the graph now have "joy full programming model" is that a missing removal or intentional?

Reply:
The graphic has "joyful programming model", which is intentional. "Developer Joy" includes a whole bunch of things, which would be implemented at different times - so what the graphic is talking about is specifically the boilerplate reduction, and not things like support for live reload or unified configuration.

Reply:
ack - I like the use of the developer joy icon. its a good fit.

staircase vs graph vs something else is for sure tricky.

Its fine and at least sends a message.

Only crazy thought I had is to make it a circle as in iteratively keep improving them but it probably gets too big :)

I'm not an artist so got some help from chatgpt:

![image](https://github.com/user-attachments/assets/6728e26a-607e-475d-9d4c-cb8e528e12ef)

I kind a like this and maybe that kind of diagram could be used as  a way to express what is explained in the paragraph that "This is a suggested pathway, not a minimum feature set."

Just a thought that came to me re-reading it. Does not invalidate the simplicity of the staircase.

if it resonate we can consider it in an update.

Reply:
I like the radar idea. I would have used different axis to not invent something else (and makes it less intimidating). 

Also some of them are boolean (codestart, you have it or not, how can you have half of a codestart (I have one but it does not compile?) - these should be incorporated into the "dev joy" axis.